### DISCUSSION TITLE
Surface reconstruction in pyvista

### DISCUSSION THREAD

**Discussion by sunitasaha** [pyvista/pyvista#4876] [Q&A]
I have surface points from a 3d object as attached 
![image](https://github.com/pyvista/pyvista/assets/70508935/a7c3a5ac-44b2-411a-9f3f-349a0ec05e65)
I have tried initially making a mesh with delaunay_3d. However I have read discussion already that it doesnt work the best. 
![image](https://github.com/pyvista/pyvista/assets/70508935/726bcef1-686a-4445-a789-6130ef178458)
Then I have tried the surface_reconstruct filter from pyvista as follow my code:
```python
point_cloud = pv.wrap(points)
final_reconsuction = point_cloud.reconstruct_surface(nbr_sz=50, progress_bar=True)
final_reconsuction = final_reconsuction.clean()
final_reconsuction = final_reconsuction.smooth(n_iter=20)
final_reconsuction.plot(show_edges=False, smooth_shading=True)
```

The results are still not a complete mesh, I also tried to do clean and smooth but no improvement. 
![image](https://github.com/pyvista/pyvista/assets/70508935/611e509b-44b8-48d9-9b5f-817f689f8375)

I tried with pymeshfix as well
```python
import pymeshfix

point_cloud = pv.PolyData(points)
surf = point_cloud.reconstruct_surface(nbr_sz=30, sample_spacing=5)
mf = pymeshfix.MeshFix(surf)
mf.repair()
repaired = mf.mesh
pl = pv.Plotter()
pl.add_mesh(point_cloud, color='k', point_size=1)
pl.add_mesh(repaired)
pl.add_title('Reconstructed Surface', font_size=12)
pl.show()
```
![image](https://github.com/pyvista/pyvista/assets/70508935/50435635-aaf7-42ed-8ff9-b45050345c2a)

Any suggestion how can I reconstruct the complete surface. Thank you in advance.

**Comment by stla** [comment]
Have you tried a package computing the alpha-hull? This may work.

**Reply by sunitasaha** [reply]
I did try from open3d alpha shape reconstruction. However it didn work either the way it is supposed to be.

**Reply by stla** [reply]
Is it possible you share the data so that I can try?

**Reply by sunitasaha** [reply]
[surface_points.txt](https://github.com/pyvista/pyvista/files/12929618/surface_points.txt)

Here is the txt file with the points. Thank you very much.

**Reply by stla** [reply]
Ok, I'll try with CGAL.

**Reply by stla** [reply]
Poisson reconstruction:

![sunitasaha_Poisson](https://github.com/pyvista/pyvista/assets/4466543/af7d5021-89dd-451c-829a-f9486d79f154)

**Reply by stla** [reply]
Alpha-wrapping reconstruction:

![sunitasaha_AlphaWrap](https://github.com/pyvista/pyvista/assets/4466543/b91cb519-08c2-4965-9e53-03b9578f97b9)

**Reply by stla** [reply]
Do they look good?

**Reply by sunitasaha** [reply]
Can you share with me the code? The alpha wrapping looks decent enough, however I have to check with the model itself.

**Reply by stla** [reply]
I'm using a R package of mine to do that. Do you know using R?

**Reply by sunitasaha** [reply]
Nope. Ok thanks. I will see what can be done.

**Reply by stla** [reply]
There's a Python package wrapping CGAL. Maybe it provides the alpha wrapping.

**Reply by stla** [reply]
(I can send you the file of the mesh also if you want.)

**Reply by sunitasaha** [reply]
I want to just make sure the mesh generated at the end is watertight.

**Reply by stla** [reply]
I think this is guaranteed by the algorithm, if I correctly remember. See the CGAL documentation about the [alpha wrapping](https://doc.cgal.org/latest/Alpha_wrap_3/index.html).
